fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/cursors/IteratorCursorBase.java

```diff
@@ -0,0 +1,142 @@
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.Iterator;
+import java.util.NoSuchElementException;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executor;
+
+/**
+ * Base cursor for {@link IteratorCursor} and {@link AsyncIteratorCursor}.
+ * @param <T> the type of elements of the cursor
+ * @param <C> the type of {@link Iterator}
+ */
+@API(API.Status.INTERNAL)
+public class IteratorCursorBase<T, C extends Iterator<T>> implements RecordCursor<T> {
+    @Nonnull
+    protected final Executor executor;
+    @Nonnull
+    protected final C iterator;
+    protected int valuesSeen;
+
+    @Nullable
+    protected CompletableFuture<Boolean> hasNextFuture;
+    @Nullable
+    protected RecordCursorResult<T> nextResult;
+
+    // for detecting incorrect cursor usage
+    protected boolean mayGetContinuation = false;
+
+    protected IteratorCursorBase(@Nonnull Executor executor, @Nonnull C iterator) {
+        this.executor = executor;
+        this.iterator = iterator;
+        this.valuesSeen = 0;
+    }
+
+    @Nonnull
+    @Override
+    @API(API.Status.EXPERIMENTAL)
+    public CompletableFuture<RecordCursorResult<T>> onNext() {
+        return CompletableFuture.completedFuture(computeNextResult(iterator.hasNext()));
+    }
+
+    protected RecordCursorResult<T> computeNextResult(boolean hasNext) {
+        mayGetContinuation = !hasNext;
+        if (hasNext) {
+            valuesSeen++;
+            nextResult = RecordCursorResult.withNextValue(iterator.next(), ByteArrayContinuation.fromInt(valuesSeen));
+        } else {
+            nextResult = RecordCursorResult.exhausted();
+        }
+        return nextResult;
+    }
+
+    @Nonnull
+    @Override
+    public CompletableFuture<Boolean> onHasNext() {
+        if (hasNextFuture == null) {
+            mayGetContinuation = false;
+            hasNextFuture = onNext().thenApply(RecordCursorResult::hasNext);
+        }
+        return hasNextFuture;
+    }
+
+    @Nullable
+    @Override
+    public T next() {
+        if (!hasNext()) {
+            throw new NoSuchElementException();
+        }
+        mayGetContinuation = true;
+        hasNextFuture = null;
+        return nextResult.get();
+    }
+
+    @Nullable
+    @Override
+    public byte[] getContinuation() {
+        IllegalContinuationAccessChecker.check(mayGetContinuation);
+        return nextResult.getContinuation().toBytes();
+    }
+
+    @Override
+    public NoNextReason getNoNextReason() {
+        return nextResult.getNoNextReason();
+    }
+
+    @Override
+    public void close() {
+        if (iterator instanceof AutoCloseable) {
+            try {
+                ((AutoCloseable) iterator).close();
+            } catch (RuntimeException ex) {
+                throw ex;
+            } catch (Exception ex) {
+                throw new RecordCoreException(ex.getMessage(), ex);
+            }
+        }
```

Comment:
This was the way it was in `IteratorCursor`, but should this call `MoreAsyncUtil::closeIterator` like the `KeyValueCursor` used to?

Reply:
I moved all the cases into `MoreAsyncUtil::closeIterator` and made the base class call that.